### ISSUE TITLE
Using `@trace` incorrectly in a static gen function gives confusing error message.

### ISSUE THREAD

**Issue by marcoct** [probcomp/Gen.jl#90]
```julia
using Gen

@gen (static) function foo()
    y = @trace(normal(0, 1), :x) + 1
    return y
end

load_generated_functions()

println(foo())
```

The issue is that the statement `y = @trace(normal(0, 1), :x) + 1` fails to parse as a `@trace` statement because of the `+ 1` at the end. Therefore, we attempt to parse it as a Julia statement, which succeeds (because we do not check the expression for `@trace`). Then, at run-time, the `@trace` macro gets expanded into a call to `traceat` that whose first argument is the gensym `##state#360`, which is only provided in the context of a dynamic `@gen` function, and is undefined in regular Julia code.

Some thought about how to handle macros in Julia code, and how to handle `@trace` in particular, is needed. We could disallow all macros, but that would seem overly restrictive. We could search the AST for `@trace` and throw a syntax error, but that seems like a special case and brittle.

Error message: 
```
ERROR: LoadError: UndefVarError: ##state#360 not defined
Stacktrace:
 [1] (::getfield(Main, Symbol("##3#4")))() at ./none:0
 [2] macro expansion at /home/marcoct/dev/Gen/src/static_ir/simulate.jl:87 [inlined]
 [3] simulate at /home/marcoct/dev/Gen/src/static_ir/simulate.jl:87 [inlined]
 [4] propose at /home/marcoct/dev/Gen/src/gen_fn_interface.jl:202 [inlined]
 [5] (::getfield(Main, Symbol("##StaticGenFunction_foo#364")))() at /home/marcoct/dev/Gen/src/static_ir/static_ir.jl:54
 [6] top-level scope at none:0
 [7] include at ./boot.jl:317 [inlined]
 [8] include_relative(::Module, ::String) at ./loading.jl:1044
 [9] include(::Module, ::String) at ./sysimg.jl:29
 [10] exec_options(::Base.JLOptions) at ./client.jl:231
 [11] _start() at ./client.jl:425
in expression starting at /home/marcoct/dev/Gen/test.jl:10
```

This issue was encountered by @fsaad

**Comment by ztangent** [comment]
Perhaps a less brittle approach would be to lift the restriction that `@trace` be the outermost expression on the RHS of an assignment, and instead create a new variable in the IR for every function call involving `@trace` as an argument? This would mean the following two definitions lead to the same static computation graph (up to variable renaming):

```julia
@gen (static) function foo1()
    y = @trace(normal(0, 1), :x) + 1
    return y
end

@gen (static) function foo2()
    z = @trace(normal(0, 1), :x)
    y = z+1
    return y
end
```

Theoretically this should be doable if `parse_julia_expr!` is written recursively, and calls `parse_trace_expr!` whenever it hits an `@trace` expression.

**Comment by marcoct** [comment]
Yes, something like that sounds reasonable, and would reduce the frequency with which this error is encountered. I think that would be a separate issue around extending the SML (https://github.com/probcomp/Gen/issues/213).

But this issue, about the quality of the error messages, still stands, separately from extending the SML. But part of rewriting the SML syntax would be making sure that more informative error messages are thrown.

**Comment by bzinberg** [comment]
I see this bug has the label "Priority" -- I don't think we've been treating it as high-pri, right?  I haven't been working on it (and should probably take myself off the assignee list).

**Comment by ztangent** [comment]
This is resolved by #214
